README
Net::IP::Match::Bin version 0.10
=============================

NAME
    Net::IP::Match::Bin - Perl extension for match IP addresses against Net
    ranges

SYNOPSIS
      use Net::IP::Match::Bin;

      my $ipm = Net::IP::Match::Bin->new();
      $ipm->add("10.1.1.0/25", ...);
      ...
      $ipm->add("192.168.2.128/26", ...);

      $cidr = $ipm->match_ip("192.168.2.131");

DESCRIPTION
    This module is XS implementation of matching IP addresses against Net
    ranges. Using similar method to Net::IP::Match::Regexp in storing Net
    ranges into memory. By implementing in XS C-code, and does not use
    regexp, more fast setup time and less using memory. This module is
    useful when Net ranges change often or reusing range data won't be
    suitable.

METHODS
    new()
        Create IP range object and initialize it.

    $ipm->add( $net | $arrayref | $hashref, ... )
        Add Network address (xxx.xxx.xxx.xxx/mask) into the object. mask is
        1 .. 32 CIDR mask bit value.

            $ipm->add( [ "10.1.0.0/16", "10.2.2.128/25", "192.168.0.0/27" ] );

            $ipm->add( { "10.1.0.0/17" => "Net A", "10.2.2.0/24" => "Net B",
                        "192.168.0.0/27" => "Net X" } );

        When HASH ref is used as arguments, match_ip() returns the value
        when matched. Note that CIDRs could be aggregated internaly, these
        values may be lost in aggregation.

    $ipm->add_range( $range_str, ... )
        Adds IP range into the object. $range_str would be

            $range_str = "1.2.3.4-10.20.30.41";

    $cidr = $ipm->match_ip( $ip )
        Searches matching $ip against previously setup networks. Returns
        matched Network in CIDR format (xxx.xxx.xxx.xxx/mask). or undef
        unless matched.

    $list = $ipm->list()
    @list = $ipm->list()
        Lists the current CIDR list which has been added previously. CIDRs
        are aggregated internaly and the list shows this internal data. In
        scalar context, list() returns reference to the array. On list
        context, returns array itself.

    $ipm->clean()
        Remove redundant CIDR data from the object. Note that this method
        does not free memories, so, if you would like to, generate list()
        and create new object from it.

FUNCTIONS
    match_ip($ip, $cidr1, $cidr2,...)
        function call match_ip() acts like as Net::IP::Match 's same name
        function. searches IP of first argument matches againsts others.

SEE ALSO
    Net::IP::Match::Regexp, Net::CIDR::Lite, Net::IP::Match

AUTHOR
    Tomo.M <tomo at cpan org>

COPYRIGHT AND LICENSE
    Copyright (C) 2013 by Tomo

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself, either Perl version 5.8.8 or, at
    your option, any later version of Perl 5 you may have available.


Changes
Revision history for Perl extension Net-IP-Match-Bin.

0.01  Wed Sep 30 22:24:45 2009
	- original version; created by h2xs 1.23 with options
		-A -n Net-IP-Match-XS2

0.01  Tue Oct 20 22:00:00 2009
	- original version; created from Net-IP-Match-XS2

0.02  Wed Oct 21 22:00:00 2009
	- added Perl alternative.
	- added bench mark result.

0.03  Thu Oct 22 00:00:00 2009
	- little bit speed-up Perl alt. verson.

0.04  Sat Nov 14 00:00:00 2009
	- Fix mem leak at add().
	- Fix _add() for registering ARRAY data.

0.05  Tue Nov 17 00:00:00 2009
	- Fix Makefile.PL which cannot compile XS.
	- little Fix document

0.06  Wed Dec  1 16:50:00 2010
	- Fix Bin.xs, lib/Net/IP/Match/Bin/Perl.pm for /32-net bug.
		Thanks, Oliver Paukstadt.

0.07  Thu Apr 18 01:18:00 2013
	- Fix Bin.xs for supress compiler warnings.
	- Fix Makefile.PL for silly compile checking.

0.08  Thu Apr 23 23:54:34 2013
	- Fix Bin.xs for _WIN32 compile errors.

0.09  Thu May  5 20:00:01 2013
	- Fix Bin.xs parse_ip hazardus when invalid data.
	- Add add_range(), list(), clean() methods.
	- some document fixes.

0.10  Thu May  9 12:00:01 2013
	- Fix Bin.xs for invalid net addition.
	- Fix _clean() for parformance flaw.

0.11  Thu May 14 00:00:05 2013
	- Remove ~Bin/Perl.pm pure-Perl module dependence.

0.12  Fri Sep 27 01:00:00 2013
	- now add()/add_range() returns the object itself when success.
	.. enabling usage like, $obj->add(\@nets)->list.

0.13  Fri Sep 27 21:10:00 2013
	- Fix add()/add_range() that was destructive when arguments
	.. were ARRAY.



Makefile.PL
use 5.005;
use Config qw(%Config);
use ExtUtils::MakeMaker;

my $DEFS = '';
my $OS = $^O;
if ($OS =~ /solaris/i) {
    $DEFS .= " -DSOLARIS";
} elsif ($OS =~ /linux/i) {
    $DEFS .= " -DLINUX";
} elsif ($OS =~ /cygwin/i) {
    $DEFS .= " -DCYGWIN";
}
my $pv = $] * 1000000;
$DEFS .= " -DPVER=$pv";

my $SRCS = [];
if (test_compile()) {
    $SRCS = ["Bin.c"];
}

# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    NAME              => 'Net::IP::Match::Bin',
    VERSION_FROM      => 'lib/Net/IP/Match/Bin.pm', # finds $VERSION
    C			=> $SRCS,
    PREREQ_PM         => {}, # e.g., Module::Name => 1.1
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM  => 'lib/Net/IP/Match/Bin.pm', # retrieve abstract from module
       AUTHOR         => 'Tomo <tomo_at_cpan.org>') : ()),
    LIBS              => [''], # e.g., '-lm'
    DEFINE            => "$DEFS", # e.g., '-DHAVE_SOMETHING'
    INC               => '-I.', # e.g., '-I. -I/usr/include/other'
	# Un-comment this if you add C files to link with later:
    # OBJECT            => '$(O_FILES)', # link all the C files too
);


sub test_compile {
    open(CONFTEST, ">conftest.c") or die "$!";
    print CONFTEST <<'EOT'; close(CONFTEST);
/* test program */
int
main ()
{

  ;
  return 0;
}
EOT

    my $cc_cmd = "$Config{cc}";
    my $exe = "conftest$Config{exe_ext}";
    $cc_cmd .= " -o $exe";

    my $rc;
    $rc = system("$cc_cmd conftest.c");
    if ($rc) {
	print "Can't compile test program rc=$rc\n\n";
	unlink("conftest.c", $exe, "conftest$Config{obj_ext}");
	return 0;
    }
    $rc = system("./$exe");
    unlink("conftest.c", $exe, "conftest$Config{obj_ext}");
    return 1 unless $rc;

    if ($rc > 0x80) {
	(my $cp = $rc) >>= 8;
	print "Test program exit status was $cp\n";
    }
    if ($rc & 0x80) {
	$rc &= ~0x80;
	unlink("*core") && print "Core dump deleted\n";
    }
    print "signal $rc\n" if $rc && $rc < 0x80;
    return 0;    
}


lib/Net/IP/Match/Bin.pm
package Net::IP::Match::Bin;

use 5.007;
use strict;
use warnings;

require Exporter;

our @ISA = qw(Exporter);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use Net::IP::Match::Bin ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(
	
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(
	match_ip
);

our $VERSION = '0.13';

require XSLoader;
XSLoader::load('Net::IP::Match::Bin', $VERSION);

# Preloaded methods go here.


1;
__END__
# Below is stub documentation for your module. You'd better edit it!

=head1 NAME

Net::IP::Match::Bin - Perl extension for match IP addresses against Net ranges

=head1 SYNOPSIS

  use Net::IP::Match::Bin;

  my $ipm = Net::IP::Match::Bin->new();
  $ipm->add("10.1.1.0/25", ...);
  ...
  $ipm->add("192.168.2.128/26", ...);

  $cidr = $ipm->match_ip("192.168.2.131");


=head1 DESCRIPTION

This module is XS implementation of matching IP addresses against Net ranges.
Using similar method to Net::IP::Match::Regexp in storing Net ranges into
memory. By implementing in XS C-code, and does not use regexp, more fast setup
time and less using memory.
This module is useful when Net ranges change often or reusing range data
won't be suitable.


=head1 METHODS

=over

=cut

=item new()

Create IP range object and initialize it.


=item $ipm->add( $net | $arrayref | $hashref, ... )

Add Network address (xxx.xxx.xxx.xxx/mask) into the object. mask is 1 .. 32 CIDR mask bit value.

    $ipm->add( [ "10.1.0.0/16", "10.2.2.128/25", "192.168.0.0/27" ] );

    $ipm->add( { "10.1.0.0/17" => "Net A", "10.2.2.0/24" => "Net B",
		"192.168.0.0/27" => "Net X" } );

When HASH ref is used as arguments, match_ip() returns the value when matched. Note that
CIDRs could be aggregated internaly, these values may be lost in aggregation.


=item $ipm->add_range( $range_str, ... )

Adds IP range into the object. $range_str would be

    $range_str = "1.2.3.4-10.20.30.41";

=item $cidr = $ipm->match_ip( $ip )

Searches matching $ip against previously setup networks. Returns matched
Network in CIDR format (xxx.xxx.xxx.xxx/mask). or undef unless matched.

=item $list = $ipm->list()

=item @list = $ipm->list()

Lists the current CIDR list which has been added previously. CIDRs are aggregated internaly and
the list shows this internal data. In scalar context, list() returns reference to the array. On
list context, returns array itself.

=item $ipm->clean()

Remove redundant CIDR data from the object. Note that this method does not free memories, so,
if you would like to, generate list() and create new object from it.

=back

=head1 FUNCTIONS

=over

=cut

=item match_ip($ip, $cidr1, $cidr2,...)

function call match_ip() acts like as Net::IP::Match 's same name function. searches IP of first
argument matches againsts others.
 

=back



=head1 SEE ALSO

=over

=item L<Net::IP::Match::Regexp>, L<Net::CIDR::Lite>, L<Net::IP::Match>

=back


=head1 AUTHOR

Tomo.M E<lt>tomo at cpan orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2013 by Tomo

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.


=cut
